### PR TITLE
Ensure that license URL is not blank

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Fixes WordPress/openverse-frontend#1656 by @sarayourfriend 

## Description

This PR ensures that for `Hits` without a `license_url`, the `license_url` field is not `None` but rather generated using the `license` and `license_version` information.

## Testing Instructions

The PR adds a unit test that covers this so passing CI should be adequate.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by obulat — CHANGES_REQUESTED** [review]
When do the hits have no `license_url`? Will a full re-indexing add `license_url` to all of the media?

**Comment by dhruvkb** [comment]
It should be fixed by a data refresh _if_ the catalog has done a backfilling of the missing field in ES.

**Comment by dhruvkb** [comment]
The API does a solid job of ensuring that the license URL information is always sent for `Media` instances. But in the search results page, the instances are of the `Hit` class which does not have the `license_url` property. For `Hit` instances, the field either exists in ES (if added during the data refresh from the catalog) and is sent in the response or doesn't exist and is sent as `null`. This PR aims to fix that and ensure it's computed for the `Hit` instances too.

Ideally a full refresh from the catalog would populate the field on existing ES documents, and this code will never be executed. But for these situations to not repeat, the catalog needs to make some guarantees about what data will always exist (so we can remove the workarounds/computations for them) and what data might not exist (so we can handle it in the API).

**Comment by sarayourfriend** [comment]
We do pass the Hits through a serializer that could pretty easily handle this kind of data transformation, for example in `to_representation`: https://github.com/WordPress/openverse-api/blob/db6f6ef24961cb0f821058d94e70b3dcc37ed70e/api/catalog/api/views/media_views.py#L96

File: api/test/unit/serializers/audio_serializers_test.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import uuid
+from unittest.mock import MagicMock
+
+from rest_framework.request import Request
+from rest_framework.test import APIRequestFactory
+
+import pytest
+
+from catalog.api.serializers.audio_serializers import AudioSerializer
```
Comment (sarayourfriend) [inline api/test/unit/serializers/audio_serializers_test.py:9]: Why is this only tested against the AudioSerializer?

Could we write parameterised tests for this that automatically test this behaviour against all subclasses?

File: api/test/unit/serializers/audio_serializers_test.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+@pytest.fixture
+def req():
+    factory = APIRequestFactory()
+    request = factory.get("/")
+    request = Request(request)
+    return request
+
+
+@pytest.fixture
+@pytest.mark.django_db
```
Comment (sarayourfriend) [inline api/test/unit/serializers/audio_serializers_test.py:21]: Is this mark necessary?

```suggestion
```

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
Change looks good to me, but if we can write the tests against all subclasses of the MediaSerializer that would be ideal.

File: api/test/unit/serializers/audio_serializers_test.py (hunk 1 above)
Comment (dhruvkb) [reply api/test/unit/serializers/audio_serializers_test.py]: That's a good idea. Implemented in 9e5a882.

File: api/test/unit/serializers/audio_serializers_test.py (hunk 2 above)
Comment (dhruvkb) [reply api/test/unit/serializers/audio_serializers_test.py]: I added this when the test used a Django instances and forgot to remove it. Removed in 8c4fd4b.

File: api/test/unit/serializers/media_serializers_test.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    )
+    return hit
+
+
+@pytest.mark.parametrize(
+    "serializer_class",
+    [
+        AudioSerializer,
+        ImageSerializer,
+    ],
```
Comment (sarayourfriend) [inline api/test/unit/serializers/media_serializers_test.py:36]: FWIW it's possible to use `__subclasses__` for this so it automatically updates itself when we add new serializers in the future.

File: api/test/unit/serializers/media_serializers_test.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+)
+def test_audio_serializer_adds_license_url_if_missing(req, hit, serializer_class):
```
Comment (sarayourfriend) [inline api/test/unit/serializers/media_serializers_test.py:38]:
```suggestion
def test_media_serializer_adds_license_url_if_missing(req, hit, serializer_class):
```

File: api/test/unit/serializers/media_serializers_test.py (hunk 3 above)
Comment (dhruvkb) [reply api/test/unit/serializers/media_serializers_test.py:36]: Subclasses are only registered when the serializers are imported. So as I see it, you'll still need to import the serializer files for the side-effect of registering subclasses.

Screenshot 2022-08-19 at 8 50 59 AM

**Review by obulat — APPROVED** [review]
Nice fix! When do we think this will be deployed?
I think we can drop the frontend change because API is where it should be fixed.

File: api/test/unit/serializers/media_serializers_test.py (hunk 3 above)
Comment (sarayourfriend) [reply api/test/unit/serializers/media_serializers_test.py:36]: Ahhh, they're not in the same modules. Rats!

**Pull request merged**